torch/csrc/autograd/custom_function.cpp

```diff
@@ -100,6 +100,33 @@ variable_list _wrap_outputs(const variable_list &input_vars,
   return outputs;
 }
 
+void check_variable_result(const Variable& original, const Variable& result, std::string hook_name) {
+  if (original.type() != result.type()) {
+    std::stringstream ss;
+    ss << "hook '" << hook_name << "' has changed the type of value (";
+    ss << "was " << original.toString() << " got ";
+    ss << result.toString() << ")";
+    throw std::runtime_error(ss.str());
+  }
+
+  if (original.is_cuda() != result.is_cuda()) {
```

Comment:
Hmm... where did you get this code from :o) Let's compare `device()` instead; that'll be more robust to non-CUDA/CPU things.

Reply:
Check that the `dtype()` is the same, that the `device()` is the same and the `layout()` is the same.

Reply:
This was just moved from python_hook.cpp

Reply:
OK. We can be semantics preserving for now, but that code should get fixed :) (File a bug, mayhap?)